File: app/src/main/java/com/qimian233/ztool/utils/FileManager.java
package com.qimian233.ztool.utils;

import android.content.ContentResolver;
import android.content.Context;
import android.net.Uri;
import android.util.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Android API 33+ 文件管理工具类
 * 支持 SAF (Storage Access Framework) 和 MediaStore 方式读写文件
 */
public class FileManager {
    private static final String TAG = "FileManager";

    /**
     * 使用 SAF 创建文件并保存配置
     */
    public static boolean saveConfigWithSAF(Context context, Uri uri, String fileName, String configContent) {
        ContentResolver resolver = context.getContentResolver();
        if (uri != null) {
            try (OutputStream outputStream = resolver.openOutputStream(uri)){
                if (outputStream != null) {
                    outputStream.write(configContent.getBytes(StandardCharsets.UTF_8));
                    outputStream.flush();
                    Log.i(TAG, "配置已保存到" + uri + fileName);
                    return true;
                } else {
                    Log.e(TAG, "输出流为空");
                    return false;
                }
            } catch (IOException e) {
                Log.e(TAG, "保存文件失败: " + e.getMessage());
                return false;
            }
        } else {
                Log.e(TAG,"uri为空");
                return false;
        }
    }

    /**
     * 使用 SAF 打开并读取文件
     */
    public static String readConfigWithSAF(Context context, Uri uri) {
        try {
            ContentResolver resolver = context.getContentResolver();
            InputStream inputStream = resolver.openInputStream(uri);
            if (inputStream != null) {
                BufferedReader reader = new BufferedReader(
                        new InputStreamReader(inputStream, StandardCharsets.UTF_8));
                StringBuilder stringBuilder = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    stringBuilder.append(line);
                }
                reader.close();
                inputStream.close();
                return stringBuilder.toString();
            }
        } catch (Exception e) {
            Log.e("SAF", "SAF 读取文件失败: " + e.getMessage());
        }
        return null;
    }


    /**
     * 生成备份文件名
     */
    public static String generateBackupFileName() {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault());
        String timestamp = sdf.format(new Date());
        return "ZTool_Config_Backup_" + timestamp + ".json";
    }

}

File: app/src/main/java/com/qimian233/ztool/utils/FileManager.kt
package com.qimian233.ztool.utils

import android.content.Context
import android.net.Uri
import android.util.Log
import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * File management utilities for Android API 33+
 * Supports reading and writing files via SAF (Storage Access Framework) and MediaStore
 */
object FileManager {
    private const val TAG = "FileManager"

    /**
     * Export a file via SAF
     * @param context context
     * @param uri target directory Uri
     * @param fileName file name to save
     * @param sourceFile source file to export
     * @return whether the export succeeded
     */
    fun exportFileWithSAF(context: Context, uri: Uri?, fileName: String, sourceFile: File?): Boolean {
        if (uri == null || sourceFile == null || !sourceFile.exists()) return false

        val resolver = context.contentResolver
        return try {
            FileInputStream(sourceFile).use { inputStream ->
                val outputStream = resolver.openOutputStream(uri)
                if (outputStream == null) return false
                outputStream.use { out ->
                    val buffer = ByteArray(1024)
                    var length: Int
                    while (inputStream.read(buffer).also { length = it } > 0) {
                        out.write(buffer, 0, length)
                    }
                }
            }
            Log.i(TAG, "file exported to $uri$fileName")
            true
        } catch (e: IOException) {
            Log.e(TAG, "failed to export file: " + e.message)
            false
        }
    }

    /**
     * Create a file via SAF and save config content
     */
    fun saveConfigWithSAF(context: Context, uri: Uri?, fileName: String, configContent: String?): Boolean {
        if (uri == null) {
            Log.e(TAG, "uri is null")
            return false
        }
        if (configContent == null) {
            Log.e(TAG, "config content is null")
            return false
        }
        val resolver = context.contentResolver
        return try {
            val outputStream = resolver.openOutputStream(uri)
            if (outputStream != null) {
                outputStream.use { out ->
                    out.write(configContent.toByteArray(StandardCharsets.UTF_8))
                    out.flush()
                    Log.i(TAG, "config saved to $uri$fileName")
                    true
                }
            } else {
                Log.e(TAG, "output stream is null")
                false
            }
        } catch (e: IOException) {
            Log.e(TAG, "failed to save file: " + e.message)
            false
        }
    }

    /**
     * Open and read a file via SAF
     */
    fun readConfigWithSAF(context: Context, uri: Uri): String? {
        return try {
            val resolver = context.contentResolver
            val inputStream = resolver.openInputStream(uri)
            if (inputStream != null) {
                val reader = BufferedReader(InputStreamReader(inputStream, StandardCharsets.UTF_8))
                val stringBuilder = StringBuilder()
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    stringBuilder.append(line)
                }
                reader.close()
                inputStream.close()
                stringBuilder.toString()
            } else {
                null
            }
        } catch (e: Exception) {
            Log.e("SAF", "SAF failed to read file: " + e.message)
            null
        }
    }

    /**
     * Generate a backup file name
     */
    fun generateBackupFileName(): String {
        val sdf = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault())
        val timestamp = sdf.format(Date())
        return "ZTool_Config_Backup_$timestamp.json"
    }
}
